compiler/rustc_trait_selection/src/error_reporting/traits/mod.rs

```diff
@@ -518,68 +499,124 @@ pub fn report_dyn_incompatibility<'tcx>(
         }
     }
 
+    attempt_dyn_to_enum_suggestion(tcx, trait_def_id, &*trait_str, &mut err);
+
+    err
+}
+
+/// Attempt to suggest converting the `dyn Trait` argument to an enumeration
+/// over the types that implement `Trait`.
+fn attempt_dyn_to_enum_suggestion(
+    tcx: TyCtxt<'_>,
+    trait_def_id: DefId,
+    trait_str: &str,
+    err: &mut Diag<'_>,
+) {
     let impls_of = tcx.trait_impls_of(trait_def_id);
-    let impls = if impls_of.blanket_impls().is_empty() {
-        impls_of
-            .non_blanket_impls()
-            .values()
-            .flatten()
-            .filter(|def_id| {
-                !matches!(tcx.type_of(*def_id).instantiate_identity().kind(), ty::Dynamic(..))
-            })
-            .collect::<Vec<_>>()
-    } else {
-        vec![]
+
+    // Don't suggest converting to an enum if there are any blanket impls.
+    if !impls_of.blanket_impls().is_empty() {
+        return;
+    }
+
+    let concrete_impls = {
```

Comment:
Done, though note this isn't a `filter + collect`-- it's a `collect` with the (likely) possibility of early-return. I moved away from the iterator version because I think the literal `return` out of the function is more straightforward to read than the `collect` into an `Option` + return on `None`, and I'm skeptical that the performance of this impl collection is relevant.